src/ReadyStateRenderer.js

```diff
@@ -27,34 +27,49 @@ class ReadyStateRenderer extends React.Component {
     const { element, querySubscription } = props;
 
     this.state = {
+      isInitialRender: true,
       element,
+      propsElement: element,
+      querySubscription,
+      selectionReference: querySubscription.retain(),
+      onUpdate: this.onUpdate,
     };
-
-    this.selectionReference = querySubscription.retain();
   }
 
   componentDidMount() {
     this.props.querySubscription.subscribe(this.onUpdate);
   }
 
-  componentWillReceiveProps(nextProps) {
-    const { element, querySubscription } = nextProps;
+  static getDerivedStateFromProps({ element, querySubscription }, state) {
+    if (state.isInitialRender) {
+      return { isInitialRender: false };
+    }
+
+    let nextState = null;
 
-    if (element !== this.props.element) {
-      this.setState({ element });
+    if (element !== state.propsElement) {
+      nextState = {
+        element,
+        propsElement: element,
+      };
     }
 
-    if (querySubscription !== this.props.querySubscription) {
-      this.selectionReference.dispose();
-      this.selectionReference = querySubscription.retain();
+    if (querySubscription !== state.querySubscription) {
+      state.selectionReference.dispose();
+      state.querySubscription.unsubscribe(state.onUpdate);
+
+      nextState = nextState || {};
+      nextState.querySubscription = querySubscription;
+      nextState.selectionReference = querySubscription.retain();
 
-      this.props.querySubscription.unsubscribe(this.onUpdate);
-      querySubscription.subscribe(this.onUpdate);
+      querySubscription.subscribe(state.onUpdate);
     }
+
+    return nextState;
   }
 
   componentWillUnmount() {
-    this.selectionReference.dispose();
+    this.state.selectionReference.dispose();
     this.props.querySubscription.unsubscribe(this.onUpdate);
```

Comment:
not the `state` one here?

Reply:
aren't they necessarily the same, by construction?